### PR TITLE
fix #20191 echo repr(nil) compiler crash (refc only), as well $nil

### LINKED ISSUE #20191: echo repr(nil) compiler crash (refc only)
### What happened?

`echo repr(nil)` works with arc and orc, but not refc.

refc output:
```
Error: internal error: genTypeInfoV1(tyNil)
No stack traceback available
To create a stacktrace, rerun compilation with './koch temp r <file>', see https://nim-lang.github.io/Nim/intern.html#debugging-the-compiler for details
```
orc and arc:
```
nil
```

### Nim Version

```
Nim Compiler Version 1.6.6 [Linux: amd64]
Compiled at 2022-05-05
Copyright (c) 2006-2021 by Andreas Rumpf

git hash: 0565a70eab02122ce278b98181c7d1170870865c
active boot switches: -d:release
```

### Current Standard Output Logs

_No response_

### Expected Standard Output Logs

_No response_

### Possible Solution

Have a check for a `nil` literal, or `isNil` inside of `repr`.

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by bung87**
fix #20191

### CODE REVIEW

File: lib/system/dollars.nim [hunk 1]
```diff
@@ -13,6 +13,7 @@ when not defined(nimPreviewSlimSystem):
     ## Outplace version of `addFloat`.
     result.addFloat(x)
 
+proc `$`*(x: pointer{lit}): string {.compileTime.} = "nil"
```
Comment (metagn) [inline lib/system/dollars.nim]: `typeof(nil) | typeof(nil)` has been used for this

Comment (bung87) [reply lib/system/dollars.nim]: duplicated?

Comment (metagn) [reply lib/system/dollars.nim]: Sorry, I mean that is how the `nil` literal has been overloaded in the past, as an alternative to `pointer{lit}`.

Comment (bung87) [reply lib/system/dollars.nim]: but why two typeof(nil) there ? one doesn't work?

Comment (metagn) [reply lib/system/dollars.nim]: Yes, using two converts it to a typeclass which is the only way it works I believe

Comment (bung87) [reply lib/system/dollars.nim]: duno which is better, my thought was user cant declare a const pointer, so `pointer{lit}` match the nil type and meaning.

File: lib/system/repr.nim [hunk 2]
```diff
@@ -12,6 +12,7 @@
 when not defined(useNimRtl):
   proc reprAny(p: pointer, typ: PNimType): string {.compilerRtl, gcsafe.}
 
+proc reprNil(): string {.compilerproc.} = return "nil"
 proc reprInt(x: int64): string {.compilerproc.} = return $x
 proc reprFloat(x: float): string {.compilerproc.} = return $x
```
Comment (konsumlamm) [inline lib/system/repr.nim]:
```suggestion
proc reprNil(): string {.compilerproc.} = "nil"
proc reprInt(x: int64): string {.compilerproc.} = $x
proc reprFloat(x: float): string {.compilerproc.} = $x
```

**Comment by Araq** [comment]
Why does `repr(nil)` compile? `nil` is not good enough as a value to guide type inference for generics.

**Comment by ringabout** [comment]
> Why does repr(nil) compile

`repr(nil)` compiles with ARC/ORC and prints 'nil' since `repr(p: pointer)` will match `nil`.

```nim
proc repr2*(p: pointer): string =
  result = "nil"
doAssert repr2(nil) == "nil"
```
The PR is not necessary for **ARC/ORC**.

**Comment by bung87** [comment]
> Why does `repr(nil)` compile? `nil` is not good enough as a value to guide type inference for generics.

remind me of this https://github.com/nim-lang/Nim/issues/18121, also nil used as optional ref type param value as well. it's un-clear to me how nil and literal nil should work.

**Pull request closed without merging**